File: Graph/Python/Count Sub Islands.py
# Problem link: https://leetcode.com/problems/count-sub-islands/description/
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        rows, cols, result = len(grid1), len(grid1[0]), 0
        visit = [[0] * cols for _ in range(rows)]

        for row in range(rows):
            for col in range(cols):
                if grid2[row][col] == 1 and visit[row][col] == 0 and self.dfs(row, col, grid1, grid2, visit):
                        result += 1
        
        return result
    
    def dfs(self, row, col, grid1, grid2, visit):
        if row < 0 or row >= len(grid1) or col < 0 or col >= len(grid1[0]) or visit[row][col] or grid2[row][col] == 0:
            return True
        
        if grid2[row][col] == 1 and grid1[row][col] == 0:
            return False
        
        visit[row][col] = 1
        
        down = self.dfs(row + 1, col, grid1, grid2, visit)
        up = self.dfs(row - 1, col, grid1, grid2, visit)
        right = self.dfs(row, col + 1, grid1, grid2, visit)
        left =  self.dfs(row, col - 1, grid1, grid2, visit)

        return left and right and down and up

File: Graph/Java/Count Sub Islands.java
// Problem link: https://leetcode.com/problems/count-sub-islands/description/
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int countSubIslands(int[][] grid1, int[][] grid2) {
        int rows = grid1.length, cols = grid1[0].length, result = 0;
        int[][] visit = new int[rows][cols];

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid2[row][col] == 1 && visit[row][col] == 0 && this.dfs(row, col, grid1, grid2, visit)) {
                    result += 1;
                } 
            }
        }

        return result;
    }

    private boolean dfs(int row, int col, int[][] grid1, int[][] grid2, int[][] visit) {
        if(row < 0 || col < 0 || row == grid1.length || col == grid1[0].length || visit[row][col] == 1 || grid2[row][col] == 0) {
            return true;
        }
        else if(grid2[row][col] == 1 && grid1[row][col] == 0) {
            return false;
        }

        visit[row][col] = 1;

        boolean up = this.dfs(row - 1, col, grid1, grid2, visit);
        boolean down = this.dfs(row + 1, col, grid1, grid2, visit);
        boolean left = this.dfs(row, col - 1, grid1, grid2, visit);
        boolean right = this.dfs(row, col + 1, grid1, grid2, visit);

        return up && down && left && right;
    }
}
